### PR TITLE
Add `QualifiedAstPath` and `PathExpr` representation

### PR DESCRIPTION
**Pull request by xFrednet**
This is a draft PR to get some early feedback on the `QualifiedAstPath` representation. [Qualified paths](https://doc.rust-lang.org/reference/paths.html?search=QualifiedPathInType#qualified-paths) can be used in [expressions](https://doc.rust-lang.org/reference/expressions/path-expr.html) and [types](https://doc.rust-lang.org/reference/types.html?highlight=QualifiedPathInType#type-expressions). These are type relative and allow the definition of the `Self` type for traits (Example: `<Vec<u32> as Default>::default()`). This can be relevant, if a struct has several functions with the same name. An example is shown at the end of the PR description.

## Representation

The current [`AstPath`](https://github.com/rust-marker/marker/blob/master/marker_api/src/ast/common/ast_path.rs#L13) representation is a simple slice of segments, which have an ident and potentially generic args. This representation is already a hybrid of [simple paths](https://doc.rust-lang.org/reference/paths.html#simple-paths) and [expression paths](https://doc.rust-lang.org/reference/paths.html#paths-in-expressions). The simple path, has the same structure as an expression path, with the difference, that it can't have generic arguments. In the current implementation, this simply means that all generic arguments are empty, for simple paths.

[Qualified paths](https://doc.rust-lang.org/reference/paths.html?search=QualifiedPathInType#qualified-paths) are not as easily combinable with the existing `AstPath` as it can contain segments with `Self` type specification. Generally, all places, where a qualified path can be used, also a normal path is usable. It therefore makes sense to combine them in some shape or form. I thought of a two main ways to represent this:

1. Everything can be represented as one `AstPath`. The segments would then be an enum, with either an identifier or the `Self` type specification.

    The disadvantage is that the segment enum, would need to be handled everywhere, even in locations where the path can never be qualified.
2. Create a wrapper type, which optionally specifies a `Self` type. This is the way rustc does it with [`QPath`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.QPath.html)

    I always found this confusing until now, when I was forces to fully understand how qualified paths work. I also think that just providing a `Path` directly, can cause users to use it directly, without considering, if the `Self` type is relevant. In most cases, this will be fine, but it might be possible to express the semantics better.

This PR proposes a variant of the second option, with a lot of documentation. The `AstPath` representation can be accessed via the `TryInto` trait or the `to_path_lossy` function. This ensures that users, have simple access to the `AstPath` representation, but don't accadently ignore potentual `Self` type specifications.

## Naming

Another question I have, is about naming. I find `QualifiedAstPath` very wordy. Can we maybe find a short version? Maybe, one of the following, if the meaning is clear
1. `QAstPath`
2. `AstQPath`
3. `QPath`
4. `QualifiedPath`

---

Example when qualified paths are important:

```rs
trait Foo {
    fn foo() {
        println!("foo() from Trait");
    }
}

struct Item;

impl Item {
    fn foo() {
        println!("foo() from Item")
    }
}
impl Foo for Item {}

pub fn main() {
    Item::foo(); // "foo() from Item"
    <Item as Foo>::foo(); // "foo() from Trait"
}
```

---

I plan to keep this as a draft PR, until I have made some modifications to the type representation and added the rustc backend, to ensure that the chosen representation can really express the semantics. The representation can still be changed later. I might create an issue in the design repo, to have a second look at the representation before making it stable.

---

cc: #52

@Niki4tap Do you maybe have any thoughts on this? :upside_down_face:

### CODE REVIEW

**Review by Niki4tap — COMMENTED** [review]
Just my personal thoughts, not sure if all of it is correct. About the naming, I like `AstQPath`, but again, that's just a personal preference

File: marker_api/src/ast/common/ast_path.rs [hunk 1]
```diff
@@ -5,15 +5,153 @@
...
+    self_ty: FfiOption<TyKind<'ast>>,
+    path_ty: FfiOption<TyKind<'ast>>,
+    path: AstPath<'ast>,
+    target: QualifiedPathTarget,
+}
+
+impl<'ast> QualifiedAstPath<'ast> {
+    /// Qualified paths can be type relative. Additionally, a type can be specified
+    /// that should be used as `Self`. The specification of the `Self` type is
+    /// sometimes needed to resolve the right item, when accessing trait implementations.
```
[the remark this answers is not in the document]
Comment (xFrednet) [reply marker_api/src/ast/common/ast_path.rs:27]: I agree, QPaths can be confusing at first. I think most users, will be fine without understanding every detail, but using the right terminology will give them the option to do further research :)

File: marker_api/src/ast/common/ast_path.rs [hunk 2]
```diff
@@ -5,15 +5,153 @@
...
 
-use super::SymbolId;
-use crate::{ast::generic::GenericArgs, context::with_cx, ffi::FfiSlice};
+use super::{Ident, ItemId, VarId};
+use crate::{
+    ast::{generic::GenericArgs, ty::TyKind},
+    ffi::{FfiOption, FfiSlice},
+};
+
+/// A path identifying a unique item. The path might be type relative
```
[the remark this answers is not in the document]
Comment (xFrednet) [reply marker_api/src/ast/common/ast_path.rs:14]: Agreed, the documentation evolved with every method, that's why it's almost non-existing at this level :)

**Comment by xFrednet** [comment]
> Just my personal thoughts, not sure if all of it is correct. About the naming, I like `AstQPath`, but again, that's just a personal preference

I wanted to know your preference, `AstQPath` is something I came up, while writing the PR and I like it. Thank you for the input, I'll make the requested changes!

**Comment by xFrednet** [comment]
Documenting the entire API will be a lot of work. Luckily, this is one of the more complex elements. I've rewritten and moved most of the docs. The changes should hopefully address all of your feedback :)

**Review by Niki4tap — APPROVED** [review]
Yep, looks good to me now!

This is still a draft, right? Not doing `bors r+` for now

**Comment by xFrednet** [comment]
Yeah a draft. I need to rework the type representation, as a path like `<<Item as Magic>::Sparkles as Firework>::fire()` could not be represented right now. I'm also not 100% sure if the current implementation covers everything. I just noticed that the path target enum was missing a variant. Though, we could merge this now and I'll continue work in followup PRs, to keep the size manageable. Do you have a preference?

**Comment by xFrednet** [comment]
Thank you for the quick reviews, I really appreciate having a second opinion!

**Comment by Niki4tap** [comment]
> Though, we could merge this now and I'll continue work in followup PRs, to keep the size manageable.
> Do you have a preference?

I don't have a preference as a reviewer, but if I was the one implementing it, I would try to implement rustc backend first, just in case :)

**Comment by xFrednet** [comment]
Oh no, I created a monster! The changes exploded *a bit*, due to the complexity of qualified paths. It sadly required a rework of the type representation. Please take your time with this review. A good amount of the changes should also be uitests.

I should almost be ready for merge, but I broke `dogfood` again

File: marker_driver_rustc/src/conversion/marker/common.rs [hunk 3]
```diff
@@ -153,10 +154,94 @@ impl<'ast, 'tcx> MarkerConversionContext<'ast, 'tcx> {
...
+                unreachable!("`to_qpath` should only be called for resolved paths, use `to_qpath_from_expr` instead")
+            },
+        }
+    }
+
+    pub fn to_qpath_from_expr(&self, qpath: &hir::QPath<'tcx>, expr: &hir::Expr<'_>) -> AstQPath<'ast> {
+        match qpath {
+            hir::QPath::Resolved(_, _) => {
+                // This one doesn't require special handling for expressions and can
+                // therefore be passed of to the normal `to_qpath`
```
Comment (Niki4tap) [inline marker_driver_rustc/src/conversion/marker/common.rs:175]:
```suggestion
                // therefore be passed off to the normal `to_qpath`
```

**Review by Niki4tap — APPROVED** [review]
A few nits, but other than that, and broken dogfood, I like it!

File: marker_api/src/ast/common/ast_path.rs [hunk 4]
```diff
@@ -5,21 +5,214 @@
...
+        target: AstPathTarget,
+    ) -> Self {
+        Self {
+            self_ty: self_ty.into(),
+            path_ty: path_ty.into(),
+            path,
+            target,
+        }
+    }
+}
```
Comment (Niki4tap) [inline marker_api/src/ast/common/ast_path.rs:183]: Shouldn't this be gated under `#[cfg(feature = "driver-api")]`?

**Comment by xFrednet** [comment]
Dogfood is sadly still a bit broken. It doesn't panic anymore, but it triggers E0220 from rustc due to the way types in items are resolved. So, the output looks something like this:

```
skipping not implemented expr at: marker_adapter/src/context.rs:70:5: 70:64 (#0)
skipping not implemented expr at: marker_adapter/src/context.rs:74:28: 74:80 (#0)
skipping not implemented expr at: marker_adapter/src/context.rs:75:5: 75:45 (#0)
    Checking marker_driver_rustc v0.1.0 (/nothing/to/see/here/marker/marker_driver_rustc)
error[E0220]: associated type `IntoIter` not found for `I`
  --> marker_driver_rustc/src/context/storage.rs:38:12
   |
38 |         I::IntoIter: ExactSizeIterator,
   |            ^^^^^^^^ there is a similarly named associated type `IntoIter` in the trait `std::iter::IntoIterator`

error[E0220]: associated type `IntoIter` not found for `I`
  --> marker_driver_rustc/src/conversion/marker.rs:76:12
   |
76 |         I::IntoIter: ExactSizeIterator,
   |            ^^^^^^^^ there is a similarly named associated type `IntoIter` in the trait `std::iter::IntoIterator`

skipping not implemented expr at: marker_driver_rustc/src/lint_pass.rs:17:38: 17:46 (#0)
skipping not implemented expr at: marker_driver_rustc/src/lint_pass.rs:23:39: 23:51 (#0)
skipping not implemented expr at: marker_driver_rustc/src/lint_pass.rs:23:53: 23:72 (#0)
```

I don't see a simple fix for this, besides updating the generic clause translation, which might be a bit much for this PR. Do you think we can leave it like this for now? Most generic clauses are translated correctly, it's just difficult, when clauses depend on each other.

**Comment by Niki4tap** [comment]
> ...updating the generic clause translation, which might be a bit much for this PR

I agree, the changes in this PR have gone way beyond the scope of it, and this seems like a good stopping point.

> Do you think we can leave it like this for now?

Sure, I don't think broken dogfood should be a blocker for merging this, as we can always fix it later.

Thank you for working on this!

bors r+

**Comment by xFrednet** [comment]
Let's see if this works

bors try

**Comment by xFrednet** [comment]
bors try

**Comment by xFrednet** [comment]
bors try

**Comment by xFrednet** [comment]
This is fun!!!!! And another try

bors try

**Comment by xFrednet** [comment]
PLEASE

bors try

**Comment by xFrednet** [comment]
I know what the problem is, but I don't know how to fix it correctly. I have another idea but not the time to test it now. That'll be a problem for tomorrow Fred

**Comment by Niki4tap** [comment]
bors try

**Comment by Niki4tap** [comment]
bors try

**Comment by Niki4tap** [comment]
bors try

**Comment by Niki4tap** [comment]
bors try

**Comment by xFrednet** [comment]
NICE! Just saw you fixing this!!

Should we r+ this?

**Comment by Niki4tap** [comment]
~~Don't worry, I got you~~

It's a shame none of us have a windows machine to test this on hand locally, I should probably virtualize it at some point, but at the time, sorry bors :)

bors r+

**Comment by xFrednet** [comment]
Thank you for fixing this!!! <3 

Out of interest, what system do you use for development?

**Comment by Niki4tap** [comment]
> Thank you for fixing this!!! <3

No problem :)

> Out of interest, what system do you use for development?

Linux. Nixos to be exact, I have [the current config](https://github.com/Niki4tap/configs) on my github, but it's not something exotic, just a simple machine with a bunch of stuff that I use daily preinstalled

**Comment by xFrednet** [comment]
Nice, we're lucky that GH provides windows and Mac machines, they should work well enough for now. :)

Improving the UI Test setup is also on my todo list, though less important and more a 1.0 milestone.

**Pull request merged**